kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableSchemaEvolutionSuite.scala

```diff
@@ -0,0 +1,1082 @@
+          "a",
+          StringType.STRING,
+          true,
+          fieldMetadataForColumn(1, "not-preserving-physical-column"))
+        .add("c", LongType.LONG, true, currentSchema.get("c").getMetadata)
+
+      assertSchemaEvolutionFails[IllegalArgumentException](
+        table,
+        engine,
+        newSchema,
+        "Existing field with id 1 in current schema has physical name")
+    }
+  }
+
+  test("Updating schema if map column nested ids are not preserved fails") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val initialSchema = new StructType()
+        .add("a", IntegerType.INTEGER, true)
+        .add(
+          "c",
+          new MapType(
+            StringType.STRING,
+            StringType.STRING,
+            false),
+          true)
+
+      createTable(engine, tablePath, initialSchema)
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val mapPhysicalColumnName =
+        currentSchema.get("c").getMetadata.getString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY)
+      val newSchema = new StructType()
+        .add("a", IntegerType.INTEGER, true, currentSchema.get("a").getMetadata)
+        .add(
+          "c",
+          new MapType(
+            StringType.STRING,
+            StringType.STRING,
+            false),
+          true,
+          fieldMetadataForMapColumn(2, mapPhysicalColumnName, "c", 5, 6))
+
+      assertSchemaEvolutionFails[IllegalArgumentException](
+        table,
+        engine,
+        newSchema,
+        "Expected field with id 2 to have nested key")
+    }
+  }
+
+  test("Updating schema and tightening nullability on existing field fails") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val initialSchema = new StructType()
+        .add("a", IntegerType.INTEGER, true)
+
+      createTable(engine, tablePath, initialSchema)
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("renamed_a", IntegerType.INTEGER, false, currentSchema.get("a").getMetadata)
+
+      assertSchemaEvolutionFails[IllegalArgumentException](
+        table,
+        engine,
+        newSchema,
+        "Cannot tighten the nullability of existing field a")
+    }
+  }
+
+  test("Cannot tighten nullability on array element") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val initialSchema = new StructType()
+        .add("a", IntegerType.INTEGER, true)
+        .add(
+          "arr",
+          new ArrayType(StringType.STRING, true))
+
+      createTable(engine, tablePath, initialSchema)
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("a", IntegerType.INTEGER, true, currentSchema.get("a").getMetadata)
+        .add(
+          "some_renamed_array",
```

Comment:
@allisonport-db I did a rename here intentionally, this would've caught the issue you pointed out where type comparison/nullability was not using the IDs.

Reply:
Can you rename the test name please then? There is no way for a reader to know that's what you are testing in this test case

Reply:
Or is this even what it's testing? I think this should be two different test cases then; we can't test two types of errors in one test case. (This is another good reason for unit tests)

Reply:
Not sure I follow, we are still testing that nullability cannot be tightened? That's the primary goal of the test. I can rename it to "Cannot tighten nullability on renamed array element" if it makes it more clear.